src/ray/core_worker/context_internal.h

```diff
@@ -0,0 +1,135 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+namespace ray {
+namespace core {
+
+class WorkerExtraContext {
+ public:
+  explicit WorkerExtraContext(const JobID &job_id)
+      : current_task_id_(), task_index_(0), put_counter_(0) {
+    SetCurrentTaskId(TaskID::FromRandom(job_id), /*attempt_number=*/0);
+  }
+
+  uint64_t GetNextTaskIndex() { return ++task_index_; }
+
+  uint64_t GetTaskIndex() const { return task_index_; }
+
+  /// Returns the next put object index. The index starts at the number of
+  /// return values for the current task in order to keep the put indices from
+  /// conflicting with return object indices. 1 <= idx <= NumReturns() is reserved for
+  /// return objects, while idx > NumReturns is available for put objects.
+  ObjectIDIndexType GetNextPutIndex() {
+    // If current_task_ is nullptr, we assume that we're in the event loop thread and
+    // are executing async tasks; in this case, we're using a fake, random task ID
+    // for put objects, so there's no risk of creating put object IDs that conflict with
+    // return object IDs (none of the latter are created). The put counter will never
+    // reset and will therefore continue to increment for the lifetime of the event
+    // loop thread (ResetCurrentTask and SetCurrenTask will never be called in the
+    // thread), so there's no risk of conflicting put object IDs, either.
+    // See https://github.com/ray-project/ray/issues/10324 for further details.
```

Comment:
It would be great if we could eventually fix this for for async actor tasks, it seems like it might be a bit easier to fix with this refactor. https://github.com/ray-project/ray/issues/10324